### ISSUE TITLE
#281 looks like it is still there in 1.1.8

### ISSUE THREAD

**Issue by miketempleman** [rueckstiess/mtools#372]
I have been trying to parse log files containing non-standard lines (text from tweets under certain conditions) and keep encountering the list index out of range:
```
 [                                        ] 0.0 % Traceback (most recent call last):
  File "/usr/local/bin/mloginfo", line 9, in <module>
    load_entry_point('mtools==1.1.8', 'console_scripts', 'mloginfo')()
  File "/usr/local/lib/python2.7/site-packages/mtools/mloginfo/mloginfo.py", line 82, in main
    tool.run()
  File "/usr/local/lib/python2.7/site-packages/mtools/mloginfo/mloginfo.py", line 77, in run
    section.run()
  File "/usr/local/lib/python2.7/site-packages/mtools/mloginfo/sections/query_section.py", line 51, in run
    for i, le in enumerate(logfile):
  File "/usr/local/lib/python2.7/site-packages/mtools/util/logfile.py", line 208, in __iter__
    le = self.next()
  File "/usr/local/lib/python2.7/site-packages/mtools/util/logfile.py", line 190, in next
    ret = le.set_datetime_hint(self._datetime_format, self._datetime_nextpos, self.year_rollover)
  File "/usr/local/lib/python2.7/site-packages/mtools/util/logevent.py", line 246, in set_datetime_hint
    if not self.split_tokens[self._datetime_nextpos-1][0].isdigit():
IndexError: list index out of range
```
There are lines in the log that are not formatted as log file lines (text with line ends) as well as being rather long. Here is a sample of such a line that seems to give set_datetime_hint in logevent.py fits:
```
2015-06-30T17:14:30.441+0000 I WRITE    [conn4072] update plover_development.socialMediaUser query: { externalID: 4207961 } update: { $set: { description: "Host of All In with Chris Hayes on MSNBC, Weeknights at 8pm. Editor at Large at The Nation. Cubs fan.
Instagram: chrislhayes
FB: https://t.co/niNbW3BZcv", URL: "http://allin.msnbc.com", profileImageURL: "http://pbs.twimg.com/profile_images/1499446228/MSNBC_Headshot_normal.jpg", createdDate: new Date(1176309838000), location: "Brooklyn", realName: "Christopher Hayes", screenName: "chrislhayes", screenNameLC: "chrislhayes", followerCount: 433760, friendCount: 796, statusesCount: 53075, urlEntities: [ "http://allin.msnbc.com" ], language: "en", verified: true, listCount: 10367, lastUpdated: new Date(1435684470206) }, $inc: { version: 1 } } nscanned:1 nscannedObjects:1 nMatched:1 nModified:1 keyUpdates:0 writeConflicts:46 numYields:1 locks:{ Global: { acquireCount: { r: 3, w: 3 } }, Database: { acquireCount: { w: 3 } }, Collection: { acquireCount: { w: 2 } }, oplog: { acquireCount: { w: 1 } } } 226ms
2015-06-30T17:14:30.441+0000 I COMMAND  [conn4072] command plover_development.$cmd command: update { update: "socialMediaUser", updates: [ { q: { externalID: 4207961 }, u: { $set: { description: "Host of All In with Chris Hayes on MSNBC, Weeknights at 8pm. Editor at Large at The Nation. Cubs fan.
Instagram: chrislhayes
FB: https://t.co/niNbW3BZcv", URL: "http://allin.msnbc.com", profileImageURL: "http://pbs.twimg.com/profile_images/1499446228/MSNBC_Headshot_normal.jpg", createdDate: new Date(1176309838000), location: "Brooklyn", realName: "Christopher Hayes", screenName: "chrislhayes", screenNameLC: "chrislhayes", followerCount: 433760, friendCount: 796, statusesCount: 53075, urlEntities: [ "http://allin.msnbc.com" ], language: "en", verified: true, listCount: 10367, lastUpdated: new Date(1435684470206) }, $inc: { version: 1 } }, multi: false, upsert: true } ], writeConcern: { w: 1 }, ordered: true, metadata: { shardName: "rsMesh", shardVersion: [ Timestamp 1318000|1, ObjectId('5275f6b9465c53a4beb395eb') ], session: 0 } } ntoreturn:1 keyUpdates:0 writeConflicts:0 numYields:0 reslen:155 locks:{ Global: { acquireCount: { r: 3, w: 3 } }, Database: { acquireCount: { w: 3 } }, Collection: { acquireCount: { w: 2 } }, oplog: { acquireCount: { w: 1 } } } 226ms
```

**Comment by miketempleman** [comment]
This awk scripts helps, but there is a bug in it when a line matches the expression /^[^0-9]|^[0-9]\D/. Since it only hammered a few log entries out of thousands I ignored it.

```
BEGIN {
    numlines=0;
    linesout=0;
}

/^[^0-9]|^[0-9]\D/ {
    lines[numlines] = $0;
    numlines++;
    next;
}

/^[0-9]{4}/ {
    for (i=0; i < numlines; i++) {
	printf " %s", lines[i];
    }
    numlines=0;
    if (linesout > 0) {
	printf "\n%s", $0;
    }
    else {
	printf "%s", $0;
    }
    linesout++;
    next;
}

END {
    for (i=0; i < numlines; i++) {
	printf " %s", lines[i];
    }
    printf "\n";
}
```

**Comment by shashankvs1** [comment]
I still see this when I am using the mloginfo command , also let me know how to use the above script . 

IndexError: list index out of range ,

QUERIES
 [==============================          ] 74.7 % Traceback (most recent call last):
  File "/usr/bin/mloginfo", line 9, in <module>
    load_entry_point('mtools==1.1.8', 'console_scripts', 'mloginfo')()
  File "/usr/lib/python2.6/site-packages/mtools/mloginfo/mloginfo.py", line 82, in main
    tool.run()
  File "/usr/lib/python2.6/site-packages/mtools/mloginfo/mloginfo.py", line 77, in run
    section.run()
  File "/usr/lib/python2.6/site-packages/mtools/mloginfo/sections/query_section.py", line 51, in run
    for i, le in enumerate(logfile):
  File "/usr/lib/python2.6/site-packages/mtools/util/logfile.py", line 208, in __iter__
    le = self.next()
  File "/usr/lib/python2.6/site-packages/mtools/util/logfile.py", line 190, in next
    ret = le.set_datetime_hint(self._datetime_format, self._datetime_nextpos, self.year_rollover)
  File "/usr/lib/python2.6/site-packages/mtools/util/logevent.py", line 246, in set_datetime_hint
    if not self.split_tokens[self._datetime_nextpos-1][0].isdigit():
IndexError: list index out of range


The version I am using is 
mtools version 1.1.8

**Comment by jasonmimick** [comment]
Running 3.0.5 where some log entries have additional "$" end-of-line characters - there are embedded in documents getting inserted, and get same error. Was able to hack source, below, but this is still an issue.

mtools/util/logevent.py line 255
            try:
                if not self.split_tokens[self._datetime_nextpos-1][0].isdigit():
                    # not the timestamp format that was hinted
                    _ = self.datetime
                    return False
                return True
            except IndexError:
                # print "Index error caught"
                return False

**Comment by rueckstiess** [comment]
Looks like this was fixed a while ago in the dev branch (#358) but hasn't made it to master yet.

I'll release version 1.1.9 shortly, that will include this fix.
